File: morphir-dotnet/src/Morphir.SDK.Core/SDK/String.fs
module Morphir.SDK.String

open Morphir.SDK
open Morphir.SDK.Maybe

let inline isEmpty string = System.String.IsNullOrEmpty(string)

let length =
    function
    | null -> 0
    | (str: string) -> str.Length

let reverse (str: string) = str |> Seq.rev |> System.String.Concat

let repeat (x: int) (str: string) = String.replicate x str

let replace (before: string) (after: string) (str: string) = str.Replace(before, after)

let append (str1: string) (str2: string) = str1 + str2

let split (sep: string) (str: string) =
    str.Split([| sep |], System.StringSplitOptions.None) |> Array.toList

let join (sep: string) (chunks: Morphir.SDK.List.List<string>) = System.String.Join(sep, chunks)

let concat (strings: string list) = join "" strings

let words (str: string) = split " " str

let lines (str: string) = split "\n" str

let slice (startIndex: int) (endIndex: int) (str: string) =
    let start =
        if startIndex >= 0 then
            startIndex
        else
            str.Length + startIndex

    let last =
        if endIndex >= 0 then
            endIndex - 1
        else
            (str.Length + endIndex - 1)

    str.[start..last]

let left (n: int) (str: string) = if n < 1 then "" else slice 0 n str

let right (n: int) (str: string) =
    if n < 1 then "" else slice -n (length str) str

let dropLeft (n: int) (str: string) =
    if n < 1 then str else slice n (length str) str

let dropRight (n: int) (str: string) = if n < 1 then str else slice 0 -n str

let contains (substring: string) (str: string) = str.Contains substring

let startsWith (substring: string) (str: string) = str.StartsWith substring

let endsWith (substring: string) (str: string) = str.EndsWith substring

let rec indexesHelp (substring: string) (str: string) (curr: int) (result: int list) =
    let idx = str.IndexOf substring

    if (idx = -1) then
        result
    else
        indexesHelp substring (slice (idx + 1) str.Length str) (curr + idx + 1) (result @ [ curr + idx ])

let indexes (substring: string) (str: string) = indexesHelp substring str 0 []

let indices (substring: string) (str: string) = indexes substring str

let toInt (str: string) =
    match System.Int32.TryParse str with
    | (true, result) -> Some(result)
    | (false, _) -> None

let fromInt (n: int) = string n

let toFloat (str: string) =
    match System.Double.TryParse str with
    | (true, result) -> Some(result)
    | (false, _) -> None

let fromFloat (n: float) = string n

let fromChar (ch: char) = string ch

let cons (ch: char) (string: string) = sprintf "%c%s" ch string

let uncons (string: string) =
    match string with
    | null -> Maybe.Nothing
    | "" -> Maybe.Nothing
    | str when (str.Length = 1) -> Maybe.Just(str.[0], System.String.Empty)
    | str -> Maybe.Just(str.[0], str.[1..])

let toList (str: string) = Seq.toList str

let fromList (chs: char list) =
    System.String.Concat(Array.ofList (chs))

let inline toUpper (str: string) = str.ToUpperInvariant()

let inline toLower (str: string) = str.ToLowerInvariant()

let pad (n: int) (ch: char) (str: string) =
    let half: float = float (n - length str) / 2.0

    repeat (ceil half |> int) (fromChar ch)
    + str
    + repeat (floor half |> int) (fromChar ch)

let padLeft (n: int) (ch: char) (str: string) =
    repeat (n - length str) (fromChar ch) + str

let padRight (n: int) (ch: char) (str: string) =
    str + repeat (n - length str) (fromChar ch)

let trim (str: string) = str.Trim()

let trimLeft (str: string) = str.TrimStart()

let trimRight (str: string) = str.TrimEnd()

let map func str = String.map func str

let filter isGood str = String.filter isGood str

let foldl (folder: char -> 'b -> 'b) (state: 'b) (str: string) : 'b =
    let mutable state = state

    for c in str do
        state <- folder c state

    state

let foldr (folder: char -> 'b -> 'b) (state: 'b) (str: string) : 'b = foldl folder state (reverse str)

let any isGood str =
    match List.tryFind isGood (toList str) with
    | None -> false
    | _ -> true

let all isGood str =
    if (List.filter isGood (toList str)).Length = str.Length then
        true
    else
        false

// Checks the exact length of a string and wraps it using the specified constructor.
let ofLength (length: int) (ctor: string -> 'a) (value: string) =
    if String.length value = length then
        Some(ctor value)
    else
        None

// Checks the max length of a string and wraps it using the specified constructor.
let ofMaxLength (maxLength: int) (ctor: string -> 'a) (value: string) =
    if String.length value <= maxLength then
        Some(ctor value)
    else
        None

let capitalize string : string =
    match uncons string with
    | Just(headChar, tailString) -> cons (Char.toUpper headChar) tailString
    | Nothing -> string

File: morphir/sdk/core/src/morphir/sdk/String.scala
/*
Copyright 2020 Morgan Stanley

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */

package morphir.sdk

import morphir.sdk.Char.Char
import morphir.sdk.Maybe.Maybe

object String {
  type String = scala.Predef.String

  @inline def isEmpty(str: String): Basics.Bool = str.isEmpty()

  @inline def length(str: String): Basics.Int = str.length()

  @inline def reverse(str: String): String = str.reverse

  def repeat(times: Basics.Int, str: String): String =
    Array.fill[String](times.toInt)(str).mkString

  @inline def replace(
    literal: String,
    replacement: String,
    target: String
  ): String =
    target.replace(literal, replacement)

  @inline def fromInt(int: Basics.Int): String =
    int.toString

  @inline def append(first: String)(second: String): String = first + second

  @inline def ++(first: String, second: String): String = first + second

  @inline def concat(strings: List[String]): String = strings.mkString

  @inline def split(sep: String)(target: String): List[String] =
    target.split(sep).toList // TODO: These aren't exactly the same

  def toInt(text: String): Maybe[Basics.Int] =
    try Maybe.just(text.toInt)
    catch {
      case _: NumberFormatException => Maybe.nothing
    }

  @inline def toUpper(text: String): String =
    text.toUpperCase()

  @inline def toLower(text: String): String =
    text.toLowerCase()

  @inline def trim(text: String): String = text.trim()

  /** Put many strings together with a given separator.
    */
  def join(sep: Char)(chunks: List[String]): String =
    chunks.mkString(sep.toString())

  /** Put many strings together with a given separator.
    */
  def join(sep: String)(chunks: List[String]): String =
    chunks.mkString(sep)

  def words(str: String): List[String] = str.split("\\s").toList

  def lines(str: String): List[String] = str.split("\\n").toList

  def slice(start: Basics.Int)(end: Basics.Int)(str: String): String =
    str.substring(start.toInt, end.toInt)

  def left(n: Basics.Int)(str: String): String = str.substring(0, n.toInt)

  def right(n: Basics.Int)(str: String): String =
    str.slice(str.length - n.toInt, str.length)

  def dropLeft(n: Basics.Int)(str: String): String = str.drop(n.toInt)

  def dropRight(n: Basics.Int)(str: String): String = str.dropRight(n.toInt)

  def contains(substring: String)(str: String): Basics.Bool = str.contains(substring)

  def startsWith(substring: String)(str: String): Basics.Bool =
    str.startsWith(substring)

  def endsWith(substring: String)(str: String): Basics.Bool = str.endsWith(substring)

  def indexes(substring: String)(str: String): List[Basics.Int] =
    str.r.findAllMatchIn(substring).map(_.start).toList

  def indices(substring: String)(str: String): List[Basics.Int] =
    indexes(substring)(str)

  def toFloat(str: String): Maybe[morphir.sdk.Float.Float] =
    try Maybe.just(str.toDouble)
    catch {
      case _: NumberFormatException => Maybe.nothing
    }

  def fromFloat(float: morphir.sdk.Float.Float): String = float.toString
  def fromFloat(float: scala.Float): String             = float.toString

  def fromChar(ch: Char): String = ch.toString

  def cons(ch: Char)(str: String): String = s"$ch$str"

  def uncons(str: String): Maybe[(Char, String)] =
    str match {
      case a if a.length == 0 => Maybe.nothing
      case a if a.length == 1 => Maybe.just((Char.from(a.charAt(0)), ""))
      case _ =>
        Maybe.just((Char.from(str.charAt(0)), str.substring(1, str.length)))
    }

  def toList(str: String): List[Char] = str.toList.map(Char.from)

  def fromList(chList: List[Char]): String = chList.mkString

  def pad(n: Basics.Int)(ch: Char)(str: String): String = {
    val padding: String = ch.toString * n.toInt
    padding + str + padding
  }

  def padLeft(n: Basics.Int)(ch: Char)(str: String): String =
    (ch.toString * n.toInt) + str

  def padRight(n: Basics.Int)(ch: Char)(str: String): String =
    str + (ch.toString * n.toInt)

  def trimLeft(str: String): String = str.replaceAll("^\\s+", "")

  def trimRight(str: String): String = str.replaceAll("\\s+$", "")

  def map(f: Char => Char)(str: String): String =
    str.toList.map(ch => f(Char.from(ch))).mkString

  def filter(f: Char => Basics.Bool)(str: String): String =
    str.toList.filter(ch => f(Char.from(ch))).mkString

  def foldl[B](f: Char => B => B)(z: B)(str: String): B =
    str.toList.foldLeft(z)((soFar, next) => f(Char.from(next))(soFar))

  def foldr[B](f: Char => B => B)(z: B)(str: String): B =
    str.toList.foldRight(z)((next, soFar) => f(Char.from(next))(soFar))

  def any(f: (Char => Basics.Bool))(str: String): Basics.Bool =
    str.toList.exists(ch => f(Char.from(ch)))

  def all(f: (Char => Basics.Bool))(str: String): Basics.Bool =
    str.toList.forall(ch => f(Char.from(ch)))

  implicit class StringOps(private val self: String) extends AnyVal {
    def ++(that: String): String = self + that
  }
}
